fix: improve dataview types

johnhooks opened the pull request:
## What?

This commit improves the types in the dataviews package.

## Why?
This is helpful by typing the `item` or `items` provided as arguments to the functions of the fields and actions of the `DataViews`.

This commit does not implement requiring these types in the `DataView` component, but helps to improve the types of the functions that are used in the `DataView` package. It also allows exporting the types to be used in consumers of the package.

## How?

- Add generic item type to all dataview types that operate on `item` or lists of `items`.

- Update the usage in the dataviews package to incorporate the use of the generic item type.

- Add an `Action` type for the dataview actions.

## Testing Instructions

Building the package types.

johnhooks committed:
fix: improve dataview types

This commit improves the types in the dataviews package.

This is helpful by typing the `item` or `items` provided as arguments
to the functions of the fields and actions of the `DataViews`.

This commit does not implement requiring these types in the `DataView`
component, but helps to improve the types of the functions that are
used in the `DataView` package. It also allows exporting the types
to be used in consumers of the package.

- Add generic item type to all dataview types that operate
on `item` or lists of `items`.

- Update the usage in the dataviews package to incorporate
the use of the generic item type.

- Add an `Action` type for the dataview actions.

johnhooks commented on packages/dataviews/src/types.ts:
```diff
@@ -32,58 +60,67 @@ export interface Field {
 	/**
 	 * The width of the field column.
 	 */
-	width: string | number | undefined;
```
`| undefined` does not mean an optional property. It's still required. `width?:` markers it as optional to the TS type checker.

johnhooks commented on packages/dataviews/src/types.ts:
```diff
@@ -32,58 +60,67 @@ export interface Field {
+	filterBy?: FilterByConfig | undefined;
 }
 
-export type NormalizedField = Required< Field >;
```
This required all properties of `Field`, but what was really intended was marking the `header`, `getValue` and `rendered` as required, as seen in the implementation of `normalize-fields.js`.

youknowriad replied:
Yes, and that's why I added `| undefined` to the others. I didn't know how to make Required just some fields like you did

johnhooks commented on packages/dataviews/src/types.ts:
```diff
@@ -164,3 +202,78 @@ export interface ViewList extends ViewBase {
+ * An action that can be performed on a record in the data set.
+ */
+export type Action< T extends Item = {} > = ActionBase< T > &
+	( WithRenderModal< T > | WithCallback< T > );
```
This appropriately types the two possibilities of providing either a `RenderModal` or `callback`. It is allows the definition of `callback` when `RenderModal` is defined, though it has a note that it will not be used (as described in the documentation).

johnhooks commented on packages/dataviews/src/view-list.tsx:
```diff
@@ -21,32 +21,27 @@ import { __ } from '@wordpress/i18n';
-	NormalizedField,
-	ViewList as ViewListType,
-} from './types';
+import type { Item, NormalizedField, ViewList as ViewListType } from './types';
```
`Data` wasn't required, as `T[]` fills it's place.

johnhooks commented on packages/dataviews/src/filter-and-sort-data-view.ts:
```diff
@@ -32,14 +32,14 @@ const EMPTY_ARRAY: Data = [];
 	if ( ! data ) {
 		return {
-			data: EMPTY_ARRAY,
+			data: EMPTY_ARRAY as T[],
```
This `as` is required because `EMPTY_ARRAY` cannot be typed generically, though is used to provide a stable object reference for an empty data array.

sirreal replied:
What's the point of this? I'm assuming it's because `EMPTY_ARRAY` is stable?

I think we can change its type to `[]` and that will satisfy the return type without requiring this assertion.

johnhooks replied:
I haven't used that before, I'll try it.

johnhooks replied:
That worked. Thanks @sirreal, I will be using that later.

johnhooks commented on packages/dataviews/src/types.ts:
```diff
@@ -3,21 +3,49 @@
 
 type Operator = 'is' | 'isNot' | 'isAny' | 'isNone' | 'isAll' | 'isNotAll';
 
 export type Item = Record< string, any >;
```
It would be nice to rename this `AnyItem`, then all the generic arguments could be `Item`, otherwise it conflicts with this declaration of `Item`.

johnhooks replied:
Prefixiing `Any` to a type used as a base for a generic argument type is a common in the `@wordpress/data` package.

youknowriad replied:
Let's do that.

johnhooks replied:
Updated. I also just captured the addition of `bulk-actions.tsx`

youknowriad commented on packages/dataviews/src/types.ts:
```diff
@@ -3,21 +3,49 @@
+/**
+ * Type of view UI.
+ */
+export type ViewType = 'table' | 'grid' | 'list';
```
I don't think we should limit the view type because we want to allow registration of views ultimately.

johnhooks replied:
I can remove it. I have just found it useful to have a type for what dataviews are available and passing a typed parameter down to the render functions of the fields. Because they typically need to render differently in specific view types.

johnhooks replied:
Changed it back to `string`.

youknowriad commented on packages/dataviews/src/normalize-fields.ts:
```diff
@@ -1,15 +1,17 @@
-export function normalizeFields( fields: Field[] ): NormalizedField[] {
+export function normalizeFields< T extends Item >(
+	fields: Field< T >[]
+): NormalizedField< T >[] {
```
I'm starting to enjoy typescript in some cases.

But this kind of changes is what makes it unbearable to a lot of people. It just makes the code way less readable. I do understand the reasoning though and don't have solutions though.

johnhooks replied:
I can't argue with you, it's not pretty. But it does facilitate deeper type checking of the items, which I have found helpful.

youknowriad commented on packages/dataviews/src/types.ts:
```diff
@@ -32,58 +60,67 @@ export interface Field {
 	 * Defaults to `item[ field.id ]`.
 	 */
-	getValue?: ( { item }: { item: Item } ) => any;
+	getValue?: ( args: { item: T } ) => string | undefined;
```
I don't think a field is necessary a string or undefined.

johnhooks replied:
There is a location in the package the return of `field.getValue` is used in a context that requires a string, I'll find it and link it here.

johnhooks replied:
Apparently after going through everything, I must have made a mistake, it's all builds with a return of `any`.

youknowriad commented on packages/dataviews/src/types.ts:
```diff
@@ -164,3 +202,78 @@ export interface ViewList extends ViewBase {
 
 export type View = ViewList | ViewBase;
+
+export interface ActionBase< T extends Item > {
```
This is the third PR that adds Action type, can we remove that from the PR for now. Just to avoid multiple PRs conflicting on the same thing.

youknowriad replied:
I think we should probably keep the PR about existing types and not add new ones. I prefer if we add new ones in their own PRs if you don't mind.

youknowriad replied:
The Action types are already on trunk, so this needs a rebase.

johnhooks replied:
@youknowriad I'll remove them and rebase.

johnhooks replied:
I'll will add the generic arguments if that's okay.

youknowriad reviewed (commented):
Thanks for taking a stab at improving the data views types.

johnhooks committed:
feat: add generic types to DataView action type

This commit expands the original PR to include generic
types for `@wordpress/dataview` actions.

This commit also does the following:

- Applies several CR suggestions from @youknowriad and
@sirreal.
- Adds generic types to to all .ts/.tsx files in the dataviews.
- Renames `Item` to `AnyItem` to match the pattern of `@wordpress/data`.
- Name all generic types to `Item` which helps to make the code more readable,
  despite the fact TypeScript generics are in general cryptic.

johnhooks committed:
add generics to bulk-actions.tsx

johnhooks committed:
replace getValue return type of any

youknowriad reviewed (approved):
Thanks for these improvements.
